<a href="https://colab.research.google.com/github/Azwadd/CS_448-Artificial-Intelligence/blob/main/Assignment%202/Assignment%202.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model

In [2]:
# Part 1. Loading the dataset
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")

# Output the first 15 rows of the data
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [3]:
# Display a summary of the table information (number of data points, etc.)
print(data.dtypes)
print(data[["Temperature °C", "Mols KCL", "Size nm^3"]].shape)
print(data["Temperature °C"].shape)
print(data["Mols KCL"].shape)
print(data["Size nm^3"].shape)
# Basically the table has 3 columns with 1000 rows each
data.info()
# Part 1. Loading the dataset - Completed

Temperature °C      int64
Mols KCL            int64
Size nm^3         float64
dtype: object
(1000, 3)
(1000,)
(1000,)
(1000,)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


In [4]:
# Part 2. Splitting the dataset
# Take the pandas dataset and split it into our features (X) and label (y)
label = np.array(data["Size nm^3"])
features = np.array(data[["Temperature °C", "Mols KCL"]])

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9)
# Part 2. Splitting the dataset - Completed

In [5]:
#Part 3. Perform a Linear Regression
# Use sklearn to train a model on the training set
lr = LinearRegression().fit(x_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
print(lr.predict(np.array([[1, 2]])))

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("Score:", lr.score(x_test, y_test))

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficients:", lr.coef_)
print("Intercept:", lr.intercept_)
#Part 3. Perform a Linear Regression - Completed - report in markdown below

[-416157.45507965]
Score: 0.8387535798248023
Coefficients: [ 870.7409581  1044.66992793]
Intercept: -419117.53589361114


$S = -419117.53589 +  870.74095(a) + 1044.66992(b) $  

The linear regression model gave us an 83.87535% accuracy. This means that the linear regression
for the data we have is somewhat accurate and there is about a 16.12465% error from the actual
data that was obtained from the series of experiments. Essentially, the values from the line
of best fit by using the linear regression model is about 16.12465% different from the
actual difference when comparing how much temperature °C and Mols KCL creates a change in
Size nm^3. This is significant, the fact that the linear regression model for our dataset
is not fully accurate is very important because it shows that we still need to improve
the training model in order to get a better line of best fit.

In [6]:
# Part 4. Use Cross Validation
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
print(cross_val_score(lr, x_test, y_test, cv=6))
# Report on their finding and their significance
# Part 4. Use Cross Validation - Completed - report in markdown below

[0.72560223 0.82278373 0.87843085 0.83540297 0.84761377 0.82112454]


The 6 cross validation scores show us that the accuracy is around 72%-88%, which shows that the
linear regression model used is accurate but not completely accurate. The 72%-88% accuracy range shows that the accuracy has a large range of 16% in between the accuracies. In fact the 72%-88% accuracy range, also shows that the model being used will not be completely accurate in any scenario that was tested. This fact that the model won't be 100% accurate is significant 
because it leads us to assume that we need to improve the accuracy. To improve the accuracy maybe we should try another model or something to help the model in order to improve the accuracy.

In [7]:
# Part 5. Using Polynomial Regression
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_train = poly.fit_transform(x_train)
X_test = poly.fit_transform(x_test)
reg = linear_model.BayesianRidge()
reg.fit(X_train, y_train)
print("Accuracy:", reg.score(X_test, y_test))
print("Intercept:", reg.intercept_)
print("Coefficients:", reg.coef_) # we get 6 coefficients
print("Predict: ", reg.predict(np.array([[1, 2, 3, 5, 6, 7]]))) # 6 predict because 6 coefficients
# Report on the metrics and output the resultant equation as you did in Part 3.
# Part 5. Using Polynomial Regression - Completed - report in markdown below

Accuracy: 1.0
Intercept: 1.7255311831831932e-05
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.37042472e-07 -2.12005968e-11
  2.00000000e+00  2.85714287e-02]
Predict:  [36.20001687]


$y = 1.72553(10^{-5}) + 0(1) + 1.2(10^2)(a) - 1.37042472e(10^{-7})(b) - 2.12005(10^{-11})(a^2) + 2(ab) + 2.85714(10^{-2})(b^2)$

The model scored 100% meaning that the polynomialfeatures with the bayesian ridge regresson model is fully accurate.
This is significant because it means that the previous model used linear regression is not capable of getting a fully accurate line of best fit by itself. The addition of polynomialfeatures gave a huge benefit and allowed the bayesian ridge model to gain enough accuracy to gain the score of 100% accuracy. This is significant because it shows that utilizing polynomialfeatures is really beneficial because we can use a multitude of degrees and fit it into a new array that is easier for the linear model to fit into. Therefore, the results show that polynomial regression is great because you can get high accuracy rates with polynomialfeatures and any compatible linear model.